### ISSUE TITLE
someone upload non-zero games time and time again on purpose

### ISSUE THREAD

**Issue by alphaladder** [gcp/leela-zero#1705]
I found someone had uploaded hundreds of same games (master vs ke Jie ) to the database,which is shameful .Any procedure to stop this evildoings.
@roy7 
![3ebfd512-522b-4d57-a375-c4638f14c15c](https://user-images.githubusercontent.com/38591796/43816639-9487f6a6-9a8a-11e8-9a67-926a64436df7.jpeg)
![066738d8-0643-40a7-b82e-4f676c9c4635](https://user-images.githubusercontent.com/38591796/43816640-94a95f94-9a8a-11e8-9fb7-7456223da42c.jpeg)

**Comment by alphaladder** [comment]
file name is c9ad259600502d664263b43cb7e542c0deabdde685e39edbde4cf51e1166eda

**Comment by storyer9999** [comment]
http://zero.sjeng.org/view/c9ad259600502d664263b43cb7e542c0deabdde685e39edbde4cf51e1166edac?viewer=wgo
help you

**Comment by alphaladder** [comment]
thanks

**Comment by betterworld** [comment]
I found this self-play game some days ago and noticed that the fuseki is very unusual for LZ. Now that I read this thread I looked up the game on ps.waltheri.net and found that it is a pro game from 1994.

http://zero.sjeng.org/view/0e9932b0115290409ddbc5bbebe7d6f4381309e48d39bdf816c67b5e175b70b3?viewer=wgo

http://ps.waltheri.net/database/game/10706/

**Comment by sbbdms** [comment]
This problem is serious...
Can it be noticed if someone uploads other kinds of manually generated demos such as totally random moves or illegal moves?

**Comment by Eddh** [comment]
Theorically it can but it would require some work.

**Comment by kuba97531** [comment]
I think there is a high chance that it's done by accident and the first step should be to make it difficult to send fake games by accident.

Right now it's very easy to do git clone, build, run and send games to the http://zero.sjeng.org/submit-match. Any modifications to either autogtp or leelaz can result in automatic sending of corrupted games.

I think both autogtp and leelaz should NOT be able to be used in official games generation / training without some deliberate action of a person building them. At the very minimum some build flag indicating that the person building the code knows what is doing.

**Comment by Ishinoshita** [comment]
Sending corrupted games by accident, because you have tweaked some parameters of LZ or autogtp before recompiling, why not. Sending, unintentionally, a replica of a 1994 pro game in training format, i.e. with visits, is much harder to imagine....

**Comment by sheeryjay** [comment]
@kuba97531 I would be really glad if this were a mistake, but can you imagine a way in which this is a mistake? People should not be modifying AutoGTP (except for disabling it perhaps), only LeelaZero. How would an unmodified AutoGTP manage to send various pro games? I hope that it only ever sends games that it initiated, it does it seems any sgf present in a directory?

Also since the metadata is correct, is it put there by LZ, or AutoGTP? If LZ then I can't imagine how it could be a pro game.

**Comment by kuba97531** [comment]
I use the classical "Never attribute to malice that which is adequately explained by ignorance/indolence".

And I can very easily imagine a way in which this is a mistake. We had already a case of games with chinese opening which was thought to be malice but turned out to be some enthusiast who wanted to recreate LZ for different set of rules.

If someone doesn't realize that the binaries built with default build settings result in uploading official matches and has made the minimal change to the LZ code so that the self-play games are some specified pro games instead of real self-play then it would result in the situation which we witness right now.

Also as I read through the code right now it looks like autoGTP is just sending back the same options it got in a request. The fact that someone "had uploaded hundreds of same games" indicates that he has changed the leelaz to replay this particular game and didn't realize him running autoGTP is constantly sending this game to the server.

**Comment by kuba97531** [comment]
I believe this can be achieved by replacing best move logic in here
https://github.com/gcp/leela-zero/blob/9e903eda6e93fb65375ea03127e4f62268913175/src/UCTSearch.cpp#L329

With something like this:
```
auto movenum = int(m_rootstate.get_movenum());
return get_master_vs_ke_jie_sgf().get_move( movenum )
```

**Comment by roy7** [comment]
We had someone a while ago sending bad data into the server, was because he was doing his own project to generate games locally with alternate rules without realizing autogtp will upload the matches to the server. If I have time at work today I'll track down the IP address(s) that uploaded these and ban them or something and put a note up specifically only to them on the main web site. That's how we alerted the prior person.

**Comment by roy7** [comment]
The two games listed above come from two different IP addresses but both reverse under the googleusercontent.com domain name. Would that suggest they are running on Colab instances or something?

The one IP has uploaded 179 games and the other 270 games since I most recently reset the log file on August 1st. In addition to getting matches/selfplay tasks and seemingly running those in normal amounts of time, there are also a couple batches of a bunch of games being uploaded with no get-task being sent. The bad game is one of the ones in that list. Someone want to look at some of these and see if they are all pro games?

8b9773e155b148d2b3176d6224737cebdccc56e7be85f9f715235a6b9d9cd5d2
d23f16e879d887ac8c891df3d75b90c4982441c9db514a3aaad08b027747870a
f92b28afde29778568bc0438f4b79a45792430b8e91826f37adfe96f73be35f3
78cadc8e48131342529d5f468cf3294ffc9081f0ddec1148da82f7bcfeb20bbf
c70a65d962e525aebc28d04f42770e8bd8596286466fde91947f2522dbfd8e5f
e4c1e5a579bc4df59404ca803530d2b97309c10e8525ae91ad262124a63b4529
65de70956e1dc7c47bc74d29aff01e590ceef1a4caade1e7072af95267f41c6b
4c5ff0217d5a1453235e2d7bb65868671067610e9915350a938696301159fbfd
466f63a9384a8eaac6d7b72c1bd50bedd928ff6b85a1296da7656fd5f960e587
a55a1a7f5642547865515943a713407aa1b31f787f6c348b8792979251082650
ba74946f37bc1a19ee4e17467356e0428e09be7da209f0726160654eff0c19a5
515dcbf348e57ece351fbe6796a2f8cb16aa50f10ea82e3654fbff2e674e9afe
5bc5261c862b65b0623623ac252942f2f97eaf7ef5368549f8fbd3252956335f
5c0a3d44c3499890002bc04e1caffbae3f72a193e114915994b7e1b7493327fd
d7dc79cf2392e9e86a81164d61bdde60e672af3556ca85f9bf3e8f08fc9741bc

Just a small random sample.

The other IP acts the same. Lots of "normal" behavior, with occasional blast of uploads of games with no get-task tied to them that I can tell.

Likely it's the same actual client, just the IP maybe changed eventually if it's a google Colab instance.

**Comment by MartinDevelopment** [comment]
Picked 4 at random, all pro games.


http://zero.sjeng.org/view/5c0a3d44c3499890002bc04e1caffbae3f72a193e114915994b7e1b7493327fd?viewer=wgo
http://ps.waltheri.net/database/game/14056/

http://zero.sjeng.org/view/5bc5261c862b65b0623623ac252942f2f97eaf7ef5368549f8fbd3252956335f?viewer=wgo
http://ps.waltheri.net/database/game/39695/

http://zero.sjeng.org/view/8b9773e155b148d2b3176d6224737cebdccc56e7be85f9f715235a6b9d9cd5d2?viewer=wgo
http://ps.waltheri.net/database/game/17444/

http://zero.sjeng.org/view/c70a65d962e525aebc28d04f42770e8bd8596286466fde91947f2522dbfd8e5f?viewer=wgo
http://ps.waltheri.net/database/game/14182/

http://zero.sjeng.org/view/65de70956e1dc7c47bc74d29aff01e590ceef1a4caade1e7072af95267f41c6b?viewer=wgo
http://ps.waltheri.net/database/game/17735/

**Comment by kennychenfs** [comment]
Or can you confirm it when uploading?

**Comment by hzyhhzy** [comment]
I think maybe the games with 60 minutes basic time were all upload by one person.But maybe now he is trying to use a more advanced way to create bad games,such as running self-plays with 100po.

**Comment by gcp** [comment]
>The two games listed above come from two different IP addresses but both reverse under the googleusercontent.com domain name. Would that suggest they are running on Colab instances or something?

Yeah, that seems likely. It's annoying to ban: either we block Colab entirely and don't get contributions there, or we have no good way to detect that in time and get some pollution (which I quite frankly don't give much about).

We might be able to file abuse reports with Google, as Colab requires a Google account.

**Comment by gcp** [comment]
>Anyway I guess now we can't say anymore LeelaZero is Zero style.

I peed in the ocean once when I was a kid, I recommend you don't go swimming and stay inside when it rains.

**Comment by gcp** [comment]
>Is it possible to require a verification code when uploading?

We do this for matches AFAIK, but for self-play there's no real way to stop this, especially if it's distributed (which Colab essentially is).

**Comment by hzyhhzy** [comment]
What about use "username"  with password?Each username has an password.Then check bad games will be easier

**Comment by hzyhhzy** [comment]
If we block Colab  , which creates nearly 70% self-play games,the bad man's goal will be achieved.

**Comment by gcp** [comment]
>If we block Colab , which creates nearly 70% self-play games,the bad man's goal will be achieved.

Is it that much? I'm not sure on the latest stats.

**Comment by gcp** [comment]
>What about use "username" with password?Each username has an password.Then check bad games will be easier

Nothing stops anyone from making hundreds of users.

**Comment by sheeryjay** [comment]
Last time usernames and tracking of users contributions was discussed, gcp was mostly against it. Partly because of voluntary willingness to follow GDPR even if it probably does not apply to the project, and partly because it is good thing to collect as few data as possible (unless one wants to get rich selling users private data).

**Comment by jokkebk** [comment]
If malicious use would become a real issue, one could theoretically run through the game using the network is was generated by and comparing the moves to 1 visit predictions (send games to random contributors for cross-checking). Most likely something like 70 % of the moves will be correct, whereas a pro game would be a lot less. Suspicious ones could be flagged.

Of course, if the user practicing this finds out the method, they can just generate 71 % using the network and rest with garbage... Only way to be really sure would be to double-generate the games. 

Having said that, I would also guess that this instance is most likely either honest mistake, or just tasteless vandalism. One might submit his own game just as a "graffiti tag", so you could later say "btw. one of those 50 million games was actually my game, I guess Leela is strong because of this".

(btw. if this was a business enterprise, letting "gold donors" submit one of their games to training archive would be a great, if untasteful, monetization method)

**Comment by theleapfaith** [comment]
Can we filter our game with a very low playouts? For example, use 50playouts to start a replay from the 10th step of a game. If this game has a large number of flags falling outside the ai selection point, we can infer that the game is maliciously mixed. I suspect that there is not only a human game in our data, but also a randomly generated game. This may be the reason we encountered problems at 20b. @gcp

**Comment by kennychenfs** [comment]
But,there is a characteristic of LeelaZero's self-play.
It is randomcnt.
How do we solution this trouble?

theleapfaith <notifications@github.com> 於 2018年8月11日 週六 12:38 寫道：

**Comment by theleapfaith** [comment]
@kennychenfs So we don't need all the same points, as long as most of them are in scope. If a game is randomly generated, we can foresee that its placement will be completely different.

**Comment by zhaoshuais** [comment]
I have a suggestion to solve this problem: we should check 1-3 moves every game by server(the cost should be 1% of our network,that's possible), as when there is no noise,the result should be same,or at least close to the best move.Then server can easily discover non-zero games.
If check 3 moves with full visits cost too much,we can even check 10 moves with smaller playouts,and check different moves with full playouts.
Later we can even send such check mission to our network,randomly,and only recheck reported games by server,then the server cost can be reduced to almost none.

**Comment by Nazgand** [comment]
Double checking moves should work if all instances use the same pseudo-random number generator. We already have the random seed info for each game in the metadata.

**Comment by jokkebk** [comment]
If it is a pseudo-random sequence and random is used in the tree search, I believe only way to double-check move X is to play it through until that? Or do we know how many random numbers were consumed by each move?

**Comment by kennychenfs** [comment]
Perhaps you can ask the user to submit a random seed.
More accurate results can then be obtained by the seed.

Joonas Pihlajamaa <notifications@github.com> 於 2018年8月13日 週一 18:09 寫道：

**Comment by gcp** [comment]
>we should check 1-3 moves every game by server(the cost should be 1% of our network,that's possible), as when there is no noise

There is always noise in self-play games.

**Comment by gcp** [comment]
The procedure described by @theleapfaith essentially works, when using cross-entropy between move selections.

**Comment by Nazgand** [comment]
We would not need to reproduce the game up to the verifying move if the calculations done on the previous move were ignored and all random things are pseudorandom with recorded seeds. The downside to this is the training would require more time, and the upside is that randomly selected moves anywhere in the game could be verified to produce the exact same training data.

Whether increasing the security through verifying games is worth the extra computation depends on how badly the fake game data hinders LeeaZero's progress.

**Comment by kennychenfs** [comment]
Even if the fake game is AlphaGoZero's game, it will lose LeelaZero's
meaning,learning from scratch.

Mark Andrew Gerads <notifications@github.com> 於 2018年8月23日 週四 19:02 寫道：

**Comment by bjiyxo** [comment]
http://zero.sjeng.org/view/1881bbe5737d6b5c4f3c1875a445bb0aff4d115bb3cbb5eee18e474ce52eb0ec?viewer=wgo

This is the Master vs Zero game 7. @gcp may delete it.

**Comment by bjiyxo** [comment]
![image](https://user-images.githubusercontent.com/30645842/46388457-c726c600-c6fe-11e8-8476-5a99d9a59817.png)

This client plays game much faster than any other clients and the games are pretty weird. 
For example,
http://zero.sjeng.org/view/4d399631703aec79e89f98066ebda624f9e56bec73c415998671324492f3b070?viewer=wgo
![image](https://user-images.githubusercontent.com/30645842/46388511-33a1c500-c6ff-11e8-9321-db85cc8aeebe.png)
http://zero.sjeng.org/view/9915b30c8d71e7493f69e5fc36a298075531cf397b1b894b7c8205e497c66645?viewer=wgo
![image](https://user-images.githubusercontent.com/30645842/46388538-592ece80-c6ff-11e8-969f-27cbbb7eb25c.png)
Many of them play to the first line or play many inefficient moves.

http://zero.sjeng.org/view/e9e5371866b7b5f7068d67144683b57ec8fe810ad9e0cf15f36825fc921064a9?viewer=wgo
![image](https://user-images.githubusercontent.com/30645842/46388610-d2c6bc80-c6ff-11e8-8f11-e13f5a714808.png)
The normal one should be like this.

@roy7 Please consider block this client temporarily.

**Comment by ainixingo** [comment]
![qq 20181003124659](https://user-images.githubusercontent.com/40348917/46390473-9e0c3280-c70a-11e8-8e69-804da600a81c.png)
![qq 20181003124656](https://user-images.githubusercontent.com/40348917/46390474-9e0c3280-c70a-11e8-9841-1c0b1d583576.png)
![qq 20181003124653](https://user-images.githubusercontent.com/40348917/46390475-9e0c3280-c70a-11e8-8676-eae61e4d4c46.png)
![qq 20181003124649](https://user-images.githubusercontent.com/40348917/46390476-9ea4c900-c70a-11e8-80a8-6ec869ad0e0c.png)
![qq 20181003124646](https://user-images.githubusercontent.com/40348917/46390477-9f3d5f80-c70a-11e8-8c8f-44c185749c88.png)
![qq 20181003124641](https://user-images.githubusercontent.com/40348917/46390478-9f3d5f80-c70a-11e8-9c7e-f7726ee1be04.png)
![qq 20181003124638](https://user-images.githubusercontent.com/40348917/46390479-9f3d5f80-c70a-11e8-9b8d-6aa584a76dfd.png)
![qq 20181003124633](https://user-images.githubusercontent.com/40348917/46390480-9fd5f600-c70a-11e8-95a9-ae19e50758c4.png)
![qq 20181003124625](https://user-images.githubusercontent.com/40348917/46390481-9fd5f600-c70a-11e8-8c75-0a33a2d5bbba.png)
![qq 20181003124711](https://user-images.githubusercontent.com/40348917/46390482-9fd5f600-c70a-11e8-8e54-b3385594f295.png)
![qq 20181003124708](https://user-images.githubusercontent.com/40348917/46390483-9fd5f600-c70a-11e8-99e0-4f2b5737db34.png)
![qq 20181003124705](https://user-images.githubusercontent.com/40348917/46390484-a06e8c80-c70a-11e8-927a-d1bdb6855bcc.png)
![qq 20181003124702](https://user-images.githubusercontent.com/40348917/46390485-a06e8c80-c70a-11e8-9a40-5086eade635d.png)

**Comment by ainixingo** [comment]
Normal should be
![n2](https://user-images.githubusercontent.com/40348917/46390516-bbd99780-c70a-11e8-9aeb-799f76bbb898.png)
![n3](https://user-images.githubusercontent.com/40348917/46390517-bc722e00-c70a-11e8-853d-89d030c47552.png)
![n1](https://user-images.githubusercontent.com/40348917/46390518-bc722e00-c70a-11e8-8a04-f0cfa78594df.png)

**Comment by seopsx** [comment]
Here's another one uploading corrupted match games:

http://zero.sjeng.org/viewmatch/43cba0db9b6de6e26655aec74466d882ccf10f5927b8c3e5dd2c9306300c3a60?viewer=wgo

Bad moves: C2, P16 (big drop in winrate)

http://zero.sjeng.org/viewmatch/09d5c9a22754eee22ac1902a855722094bb6f8b7a4f8bc740af134adb5d6fccb?viewer=wgo

Bad moves: E4, G7, F6, R3, P17

http://zero.sjeng.org/viewmatch/4a249a8b31c531c9eb599a2dc82727ecf173fcf9d1d67e5f5f08be4ed45d9f7c?viewer=wgo

![bad_games_4](https://user-images.githubusercontent.com/35587223/46394038-6ee8c780-c6f0-11e8-9c47-f87660927204.png)


Seems malicious.

**Comment by roy7** [comment]
I did check e9e5371866b7b5f7068d67144683b57ec8fe810ad9e0cf15f36825fc921064a9 but I can't check them all since I'm busy most of today at work, and that one came from Google so I assume Colab. Is it possible something about recent next changes or whatever isn't working right on Colab? The fp16 or whatever.

**Comment by Ishinoshita** [comment]
@roy7 Is there a way to linked a training sample to the corresponding self-play game hash? I'm curious to have a look at the policy distribution for such so-called malicious games.

**Comment by bjiyxo** [comment]
IMO, the easiest way to check the strange game is check the game duration. We can check the game that > 100 move/min (or something like that), and consider it as suspicious client.
